docs/design/openstack/networking-infrastructure.md

```diff
@@ -0,0 +1,71 @@
+
+## Virtual IP's
+
+We use virtual IP addresses, VIPs, managed by keepalived
+to provide high availability access to essential APIs and services. For more info
+on how this works, please read about what [keepalived is](https://www.keepalived.org/) and
+about the underlying [VRRP algorithm](https://en.wikipedia.org/wiki/Virtual_Router_Redundancy_Protocol)
+that it runs. In our current implementation, we have 3 highly available VIPs that we manage: 
+Ingress, which handles requests to services managed by OpenShift, DNS, which handles internal dns requests, and API, which handles requests to the openshift API. Our VIP addresses are chosen and validated from the nodes subnet in the openshift
+installer, however our VIPs along with our internal dns and loadbalancing is managed
+by the [machine-config-operator(MCO)](https://github.com/openshift/machine-config-operator/tree/master/docs).
+The MCO has been configured to run static pods on the Bootstrap, Master, and Worker Nodes that
+run our internal networking infrastructure. Files run on the bootstrap node can be found 
+[here](https://github.com/openshift/machine-config-operator/tree/master/manifests/openstack).
+Files run on both master and worker nodes can be found 
+[here](https://github.com/openshift/machine-config-operator/tree/master/templates/common/openstack/files).
+Files run on only master nodes can be found 
+[here](https://github.com/openshift/machine-config-operator/tree/master/templates/master/00-master/openstack/files).
+Lastly, files run only on worker nodes can be found 
+[here](https://github.com/openshift/machine-config-operator/tree/master/templates/worker/00-worker/openstack/files).
+
+## Infrastructure Walkthrough
+
+The bootstrap node is responsible for running temporary networking infrastructure while the Master
+nodes are still coming up. The bootstrap node will run a coredns instance, as well as 
+keepalived. While the bootstrap node is up, it will have priority running the API and DNS
+VIPs.
+
+The Master nodes run dhcp, haproxy, coredns, mdns publisher, and keepalived. Haproxy loadbalances incoming requests 
+to the api across all running masters. It also runs a stats and healthcheck server. Keepalived manages all 3 VIPs on the master, where each
+master has an equal chance of being assigned one of the VIPs. Be aware, the bootstrap node has the highest priority for hosting the DNS
+and API VIPs, so they will point to addresses there for as long as they can. Once the master nodes have stood up a fully functioning control plane, 
+the bootstrap node will be torn down by the installer. Keepalived implements periodic health checks to make sure
+that essential services can be reached by their VIPs, when it determines that they cannot, it alters the VIP to point to a server it deems fit.  
+To ensure the api is reachable through the API VIP, keepalived periodically attempts to reach the api through the API VIP. It will do the same 
```

Comment:
The check is not through the API VIP, but using localhost. It is HAProxy monitor who checks via the API VIP to see if the API LB is ready.

Reply:
why does it do this?

Reply:
because on each node, nothing listends to the API VIP until that is configured by keepalived. So Keepalived itself, to be able to tell if it can configure the VIP, needs to use either the non virtual IP of the node or localhost.

The note about HAProxy means that HAProxy will only put an iptables rule redirecting the API traffic to itself when it is hosting the VIP.